zcash_primitives/src/memo/structured.rs

```diff
@@ -0,0 +1,325 @@
+//! Handlers for structured memos.
+
+use wasabi_leb128::{ReadLeb128, WriteLeb128};
+
+use super::{Error, TextMemo};
+use crate::sapling::PaymentAddress;
+
+/// A payload within a [`StructuredMemo`].
+#[derive(Clone, Debug, PartialEq)]
+pub enum Payload {
+    /// A Sapling return address.
+    ReturnAddress(PaymentAddress),
```

Comment:
IIRC I copied this type code from some existing implementation of structured memos. It was Sapling-specific then, but we can easily expand its definition to be any valid recipient address encoding.

Note that this uses a byte encoding rather than a string encoding for more efficient packing, so we may need to consider how this interacts with Unified Addresses (which now exist).